can't access emitted filename within `generateBundle`

### Rollup Version

3.7.0

### Operating System (or Browser)

MacOS

### Node Version (if applicable)

v18.12.1

### Link To Reproduction

https://github.com/danielroe/rollup-hash-repro

### Expected Behaviour

I expect the filename generated to be available within `generateBundle`.

**Note**: I didn't include the REPL.it link as it worked as expected there (with rollup v1.31.0) _until_ I upgraded the rollup version to 3.7.0 upon which it failed entirely with:

```
npx rollup --version & npx rollup -c
Cannot find module 'node:process'
Require stack:
- /home/runner/rollup-hash-repro/node_modules/rollup/dist/bin/rollup
exit status 1
```

### Actual Behaviour

```
❯ pnpm build

> rollup-hash-repro@1.0.0 build /Users/daniel/sandbox/rollup-hash-repro
> rollup -c rollup.config.js


src/index → stdout...
something.mjs-!~{001}~.js

//→ index.js:
const foo = 'bar';

export { foo };

//→ something.mjs-58e57b33.js:
var test = 'bobaloo';

export { test as default };
```

--- a/src/Chunk.ts
+++ b/src/Chunk.ts
@@ -163,7 +163,6 @@ export default class Chunk {
 	execIndex: number;
 	exportMode: 'none' | 'named' | 'default' = 'named';
 	facadeModule: Module | null = null;
-	id: string | null = null;
 	namespaceVariableName = '';
 	suggestedVariableName: string;
 	variableName = '';
@@ -318,6 +317,32 @@ export default class Chunk {
 		return true;
 	}
 
+	finalizeChunk(
+		code: string,
+		map: SourceMap | null,
+		hashesByPlaceholder: Map<string, string>
+	): OutputChunk {
+		const renderedChunkInfo = this.getRenderedChunkInfo();
+		const finalize = (code: string) => replacePlaceholders(code, hashesByPlaceholder);
+		const fileName = (this.fileName = finalize(renderedChunkInfo.fileName));
+		return {
+			...renderedChunkInfo,
+			code,
+			dynamicImports: renderedChunkInfo.dynamicImports.map(finalize),
+			fileName,
+			implicitlyLoadedBefore: renderedChunkInfo.implicitlyLoadedBefore.map(finalize),
+			importedBindings: Object.fromEntries(
+				Object.entries(renderedChunkInfo.importedBindings).map(([fileName, bindings]) => [
+					finalize(fileName),
+					bindings
+				])
+			),
+			imports: renderedChunkInfo.imports.map(finalize),
+			map,
+			referencedFiles: renderedChunkInfo.referencedFiles.map(finalize)
+		};
+	}
+
 	generateExports(): void {
 		this.sortedExportNames = null;
 		const remainingExports = new Set(this.exports);
@@ -449,31 +474,6 @@ export default class Chunk {
 		return facades;
 	}
 
-	generateOutputChunk(
-		code: string,
-		map: SourceMap | null,
-		hashesByPlaceholder: Map<string, string>
-	): OutputChunk {
-		const renderedChunkInfo = this.getRenderedChunkInfo();
-		const finalize = (code: string) => replacePlaceholders(code, hashesByPlaceholder);
-		return {
-			...renderedChunkInfo,
-			code,
-			dynamicImports: renderedChunkInfo.dynamicImports.map(finalize),
-			fileName: finalize(renderedChunkInfo.fileName),
-			implicitlyLoadedBefore: renderedChunkInfo.implicitlyLoadedBefore.map(finalize),
-			importedBindings: Object.fromEntries(
-				Object.entries(renderedChunkInfo.importedBindings).map(([fileName, bindings]) => [
-					finalize(fileName),
-					bindings
-				])
-			),
-			imports: renderedChunkInfo.imports.map(finalize),
-			map,
-			referencedFiles: renderedChunkInfo.referencedFiles.map(finalize)
-		};
-	}
-
 	getChunkName(): string {
 		return (this.name ??= this.outputOptions.sanitizeFileName(this.getFallbackChunkName()));
 	}
@@ -483,7 +483,7 @@ export default class Chunk {
 	}
 
 	getFileName(): string {
-		return this.preliminaryFileName?.fileName || this.getPreliminaryFileName().fileName;
+		return this.fileName || this.getPreliminaryFileName().fileName;
 	}
 
 	getImportPath(importer: string): string {

--- a/src/utils/FileEmitter.ts
+++ b/src/utils/FileEmitter.ts
@@ -144,8 +144,7 @@ function getChunkFileName(
 		return file.fileName;
 	}
 	if (facadeChunkByModule) {
-		const chunk = facadeChunkByModule.get(file.module!)!;
-		return chunk.id || chunk.getFileName();
+		return facadeChunkByModule.get(file.module!)!.getFileName();
 	}
 	return error(errorChunkNotGeneratedForFileName(file.fileName || file.name));
 }

--- a/src/utils/renderChunks.ts
+++ b/src/utils/renderChunks.ts
@@ -309,7 +309,7 @@ function addChunksToBundle(
 			map.file = replacePlaceholders(map.file, hashesByPlaceholder);
 			updatedCode += emitSourceMapAndGetComment(finalFileName, map, pluginDriver, options);
 		}
-		bundle[finalFileName] = chunk.generateOutputChunk(updatedCode, map, hashesByPlaceholder);
+		bundle[finalFileName] = chunk.finalizeChunk(updatedCode, map, hashesByPlaceholder);
 	}
 	for (const { chunk, code, fileName, map } of nonHashedChunksWithPlaceholders) {
 		let updatedCode =
@@ -317,7 +317,7 @@ function addChunksToBundle(
 		if (map) {
 			updatedCode += emitSourceMapAndGetComment(fileName, map, pluginDriver, options);
 		}
-		bundle[fileName] = chunk.generateOutputChunk(updatedCode, map, hashesByPlaceholder);
+		bundle[fileName] = chunk.finalizeChunk(updatedCode, map, hashesByPlaceholder);
 	}
 }
 

--- a/test/function/samples/emit-chunk-hash/_config.js
+++ b/test/function/samples/emit-chunk-hash/_config.js
@@ -1,0 +1,20 @@
+const assert = require('node:assert');
+let referenceId;
+
+module.exports = {
+	description: 'gives access to the hashed filed name via this.getFileName in generateBundle',
+	options: {
+		input: 'main',
+		output: {
+			chunkFileNames: '[name]-[hash].js'
+		},
+		plugins: {
+			buildStart() {
+				referenceId = this.emitFile({ type: 'chunk', id: 'emitted' });
+			},
+			generateBundle() {
+				assert.strictEqual(this.getFileName(referenceId), 'emitted-38bdd9b2.js');
+			}
+		}
+	}
+};

--- a/test/function/samples/emit-chunk-hash/emitted.js
+++ b/test/function/samples/emit-chunk-hash/emitted.js
@@ -1,0 +1,1 @@
+assert.ok(true);

--- a/test/function/samples/emit-chunk-hash/main.js
+++ b/test/function/samples/emit-chunk-hash/main.js
@@ -1,0 +1,1 @@
+assert.ok(true);

